substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrThreadLocal.java

```diff
@@ -168,9 +177,20 @@ public static void stopRecording(IsolateThread isolateThread, boolean freeJavaBu
         assert samplerWriterData.get(isolateThread).isNull();
 
         SamplerBuffer buffer = samplerBuffer.get(isolateThread);
+
         if (buffer.isNonNull()) {
-            SubstrateJVM.getSamplerBufferPool().pushFullBuffer(buffer);
-            samplerBuffer.set(isolateThread, WordFactory.nullPointer());
+            BufferNode node = buffer.getNode();
+            // Must lock here to avoid races with a flushing thread processing buffers
```

Comment:
Flushing must not be in progress when we stop recording. Maybe we should acquire the `JfrChunkWriter` lock before we enqueue the VM operation in `SubstrateJVM.endRecording()`.

Reply:
Added the `JfrChunkWriter` lock acquisition